skill-tree /src/main/java/com/RDS/skilltree/utils/JWTUtils.java

```diff
@@ -0,0 +1,79 @@
+package com.RDS.skilltree.utils;
+
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.Jwts;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.security.authentication.AuthenticationCredentialsNotFoundException;
+import org.springframework.stereotype.Component;
+
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.interfaces.RSAPrivateKey;
+import java.security.interfaces.RSAPublicKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.Base64;
+
+@Component
+public class JWTUtils {
+
+    @Value("${jwt.rds.public.key}")
+    private String publicRDSKeyString;
+
+    /**
+     * Converts the given private key string to an RSAPrivateKey object.
+     *
+     * @param privateKeyString The private key string to be converted.
+     * @return The corresponding RSAPrivateKey object.
+     * @throws Exception If an error occurs during the conversion process.
+     */
+    private RSAPrivateKey convertToRSAPrivateKey(String privateKeyString) throws Exception {
+        try {
+            privateKeyString = privateKeyString.replace("-----BEGIN PRIVATE KEY-----", "")
+                    .replace("-----END PRIVATE KEY-----", "")
+                    .replaceAll("\\s", "");
+
+            byte[] privateKeyBytes = Base64.getDecoder().decode(privateKeyString);
+            PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(privateKeyBytes);
+            KeyFactory keyFactory = KeyFactory.getInstance("RSA");
+            return (RSAPrivateKey) keyFactory.generatePrivate(keySpec);
+        } catch (NoSuchAlgorithmException | InvalidKeySpecException | IllegalArgumentException e) {
+            throw new Exception("Error converting private key: " + e.getMessage(), e);
+        }
+    }
+
+    /**
+     * Converts the given public key string to an RSAPublicKey object.
+     *
+     * @param publicKeyString the public key string to be converted
+     * @return the RSAPublicKey object converted from the public key string
+     * @throws Exception if there is an error during the conversion process
+     */
+    private RSAPublicKey convertToRSAPublicKey(String publicKeyString) throws Exception {
```

Comment:
optimize on below points please


Use a try-with-resources block. The KeyFactory.getInstance() method throws a NoSuchAlgorithmException if the specified algorithm is not supported. To handle this exception, you can use a try-with-resources block. This will ensure that the KeyFactory object is automatically closed, even if an exception is thrown.

Use a catch block to handle specific exceptions. The KeyFactory.generatePublic() method throws an InvalidKeySpecException if the specified key spec is invalid. To handle this exception, you can use a catch block. This will allow you to return a more specific error message to the caller.

Cache the public key. The convertToRSAPublicKey() method is likely to be called multiple times with the same public key string. To improve performance, you can cache the public key so that it does not need to be generated each time.

Reply:
> Use a try-with-resources block. The KeyFactory.getInstance() method throws a NoSuchAlgorithmException if the specified algorithm is not supported. To handle this exception, you can use a try-with-resources block. This will ensure that the KeyFactory object is automatically closed, even if an exception is thrown.

https://github.com/Real-Dev-Squad/skill-tree-backend/pull/22#discussion_r1390195751

> Cache the public key. The convertToRSAPublicKey() method is likely to be called multiple times with the same public key string. To improve performance, you can cache the public key so that it does not need to be generated each time.

Tried caching this value, but the values are getting flushed from the `ConcurrentHashMap`  the reason to be debugged.

Reply:
I tried caching the public key, since this is used in the filter the values were getting refreshed, will once check this